server/src/main/java/org/apache/lucene/index/BaseSingleStarTreeBuilder.java

```diff
@@ -0,0 +1,660 @@
+
+    /**
+     * Returns the star-tree document of the given document id in the star-tree.
+     *
+     * @return Star tree document
+     */
+    public abstract List<StarTreeDocument> getStarTreeDocuments() throws IOException;
+
+    /**
+     * Returns the dimension value of the given document and dimension id in the star-tree.
+     *
+     * @param docId       Document Id
+     * @param dimensionId Dimension Id
+     * @return Dimension value
+     */
+    public abstract long getDimensionValue(int docId, int dimensionId) throws IOException;
+
+    /**
+     * Sorts and aggregates the star-tree Document in the segment, and returns a star-tree document iterator for all the
+     * aggregated star-tree document.
+     *
+     * <p>This method reads star-tree document from segment and generates the initial star-tree document for the star-tree.
+     *
+     * @param numDocs Number of documents in the segment
+     * @return Iterator for the aggregated star-tree document
+     */
+    public abstract Iterator<StarTreeDocument> processSegmentStarTreeDocuments(int numDocs) throws IOException;
+
+    /**
+     * Generates aggregated star-tree document for star-node.
+     *
+     * <p>This method will do the following steps:
+     *
+     * <ul>
+     *   <li>Creates a temporary buffer for the given range of documents
+     *   <li>Replaces the value for the given dimension Id to {@code STAR}
+     *   <li>Sorts the starTreeDocument inside the temporary buffer
+     *   <li>Aggregates the starTreeDocument with same dimensions
+     *   <li>Returns an iterator for the aggregated starTreeDocument
+     * </ul>
+     *
+     * @param startDocId  Start document id in the star-tree
+     * @param endDocId    End document id (exclusive) in the star-tree
+     * @param dimensionId Dimension id of the star-node
+     * @return Iterator for the aggregated starTreeDocument
+     */
+    public abstract Iterator<StarTreeDocument> generateStarTreeForStarNode(int startDocId, int endDocId, int dimensionId)
+        throws IOException;
+
+    /**
+     * Returns the next segment star-tree document
+     */
+    protected StarTreeDocument getNextSegmentStarTreeDocument() throws IOException {
+        long[] dimensions = getNextSegmentStarTreeDocumentDimensions();
+        Object[] metrics = getNextSegmentStarTreeDocumentMetrics();
+        return new StarTreeDocument(dimensions, metrics);
+    }
+
+    /**
+     * Returns the next segment star-tree document for the dimensions
+     *
+     * @return dimension values for each of the star-tree dimension
+     * @throws IOException when we are unable to iterate to the next doc
+     */
+    long[] getNextSegmentStarTreeDocumentDimensions() throws IOException {
+        long[] dimensions = new long[numDimensions];
+        for (int i = 0; i < numDimensions; i++) {
+            try {
+                dimensionReaders[i].nextDoc();
```

Comment:
Need end of reader checks and better error handling , lets throw exception . Also how are we handling NULL values here ?

Reply:
ack!